File: backend/schemas/budget.py
"""Budget schemas (Pydantic v2): editable monthly lines + budget-vs-actual."""

from pydantic import BaseModel, Field


class BudgetLineIn(BaseModel):
    category: str = Field(min_length=1)
    account_prefix: str = Field(min_length=1)
    months: list[float] = Field(default_factory=lambda: [0.0] * 12)  # Jan..Dec


class BudgetSaveIn(BaseModel):
    fiscal_year: str = Field(min_length=1)
    lines: list[BudgetLineIn] = []


class BudgetLineRead(BaseModel):
    category: str
    account_prefix: str
    kind: str                       # "income" | "expense"
    months_budget: list[float]      # 12 planned amounts
    months_actual: list[float]      # 12 actuals from the ledger
    budget: float                   # yearly total budget
    actual: float                   # yearly total actual
    variance: float                 # budget − actual
    pct: float                      # actual as % of budget


class BudgetReport(BaseModel):
    fiscal_year: str
    lines: list[BudgetLineRead]
    total_budget: float
    total_actual: float

File: frontend/lib/budget.ts
import { api } from "@/lib/http";

/** Curated SYSCOHADA account groups a budget line can track against. */
export const ACCOUNT_GROUPS: { prefix: string; label: string }[] = [
  { prefix: "70", label: "Revenue (sales)" },
  { prefix: "601", label: "Purchases — goods" },
  { prefix: "61", label: "Transport" },
  { prefix: "62", label: "External services & rent" },
  { prefix: "64", label: "Taxes & duties" },
  { prefix: "66", label: "Salaries & staff" },
  { prefix: "67", label: "Financial charges" },
  { prefix: "68", label: "Depreciation & provisions" },
  { prefix: "65", label: "Other expenses" },
];

export type BudgetLine = {
  category: string;
  account_prefix: string;
  kind: "income" | "expense";
  months_budget: number[]; // 12
  months_actual: number[]; // 12
  budget: number;          // year total
  actual: number;          // year total
  variance: number;
  pct: number;
};
export type BudgetReport = {
  fiscal_year: string;
  lines: BudgetLine[];
  total_budget: number;
  total_actual: number;
};
export type BudgetSaveLine = { category: string; account_prefix: string; months: number[] };

export const getBudget = (token: string, fiscalYear: string) =>
  api.get<BudgetReport>(token, "/budget", { fiscal_year: fiscalYear });

export const saveBudget = (
  token: string,
  input: { fiscal_year: string; lines: BudgetSaveLine[] }
) => api.put<BudgetReport>(token, "/budget", input);
